<?xml version="1.0"?>

<robot name="gazebo_M.A.R.S_prototype" xmlns:xacro="http://www.ros.org/wiki/xacro">
<xacro:property name="radius" value="0.075" />
<xacro:property name="length" value="0.06" />

<!-- ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////-->
<!--MATERIAL COLORS -->
  <material name="blue">
    <color rgba="0 0 1 1"/>
  </material>

<material name="red">
    <color rgba="1 0 0 1"/>
  </material>


<material name="pink">
    <color rgba="0.8 0 1 0.5"/>
  </material>


    <material name="white">
    <color rgba="1 1 1 1"/>
  </material>

  <material name="green">
    <color rgba="0 1 0 1"/>
</material>

<gazebo reference="base_link">
  <material>Gazebo/orange</material>
</gazebo>

<!--/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////-->
<!--/////////////////////// MACROS FOR GAZEBO TRANSMISSION TAGS AND ROS CONTROL PLUGIN ///////////////////////////////////////////////-->
 <!-- plugin -->
 
 <xacro:macro name="transmission_block" params="joint_name">
	    <transmission name="joint_tranmission">
	    <type>transmission_interface/SimpleTransmission</type>
	    <joint name="${joint_name}">
	        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
	    </joint>
	    <actuator name="joint_motor">
	        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
	        <mechanicalReduction>1</mechanicalReduction>
	    </actuator>
	  </transmission>
   </xacro:macro>

<!--///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////-->
<!-- Macros for inertia matrix calculations-->
<property name="M_PI" value="3.1415926535897931" />
 <xacro:macro name="sphere_inertial" params="radius mass">
    <inertial>
      <mass value="${mass}" />
      <origin xyz="0 0 0" />
      <inertia ixx="${0.4 * mass * radius * radius}" ixy="0.0" ixz="0.0"
        iyy="${0.4 * mass * radius * radius}" iyz="0.0"
        izz="${0.4 * mass * radius * radius}" />
    </inertial>
  </xacro:macro>

  <xacro:macro name="cylinder_inertial" params="radius length mass x y z">
    <inertial>
      <mass value="${mass}" />
      <origin xyz="${x} ${y} ${z}" />
      <inertia ixx="${0.0833333 * mass * (3 * radius * radius + length * length)}" ixy="0.0" ixz="0.0"
        iyy="${0.0833333 * mass * (3 * radius * radius + length * length)}" iyz="0.0"
        izz="${0.5 * mass * radius * radius}" />
    </inertial>
  </xacro:macro>

<!--/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////-->
<!-- MOBILE PLATFORM CHASSIS-->
  <link name="base_link">
      <origin rpy="0 0 0" xyz="0 0 0"/>
 </link>
<joint name="dummy_link_to_base_link" type="fixed">
  <axis xyz="0 0 0"/> 
<origin rpy="0 0 0" xyz="0 0 0"/>
 <!-- <limit effort="100" velocity="100"/> -->
    <parent link="base_link"/> 
    <child link="dummy_link_chassis"/>
        
   
  </joint>
<!--/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////-->

<!--Visual of the Link-->
  <link name="dummy_link_chassis">
    <visual>
    
    <geometry>
  
        <box size="0.5 0.40 0.40"/>
       </geometry>
  <!-- Setting up the origin of the link -->
      <origin rpy="0 0 0" xyz="0 0 0.275"/>
       <material name="blue"/>
    </visual>
<!--  COLLISION -->
    <collision>
      <geometry>
       <box size="0.51 0.41 0.41"/>
       
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0.275"/>
    </collision>
<!--  INERTIAL -->
    <inertial>
      <mass value="5.0"/>
      
      <inertia
          ixx="0.1333333" ixy="0" ixz="0"
          iyy="0.1708333" iyz="0"
          izz="0.1708333"
      />
      <origin rpy="0 0 0" xyz="0 0 0.275"/>
    </inertial>
  
  </link>
 
 
<!--/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////-->
<!--WHEEL_1--> 
<!--Visual of the Link-->
  <link name="wheel1">
    <visual>
      <geometry>
      <cylinder radius="${radius}" length="${length}"/>
      </geometry>
  <!-- Setting up the origin of the link -->
       <origin rpy="0 1.57075 1.57075" xyz="0 0 0"/>
        <material name="white"/>
    </visual>
     <collision>
<!--  COLLISION -->
     <origin rpy="0 1.57075 1.57075" xyz="0 0 0"/>
      <geometry>
         <cylinder radius="${radius}" length="${length}"/>
      </geometry>
    </collision>
<!--  INERTIAL -->
 <inertial>
      <origin xyz="0 0 0" rpy="0 1.5707 1.5707"/>
      <mass value="0.4"/>
      
      <inertia
        ixx="0.0006825" ixy="0.0" ixz="0.0"
        iyy="0.0006825" iyz="0.0"
        izz="0.001125"/>
    </inertial>

  </link>

<joint name="base_link_to_wheel1" type="continuous">
  <axis xyz="0 1 0"/> 
<origin rpy="0 0 0" xyz="0.125 0.23 0.075"/>
 <!-- <limit effort="100" velocity="100"/> -->
 <joint_properties damping="0.0" friction="0.0"/>
    <parent link="base_link"/> 
    <child link="wheel1"/>
        
   
  </joint>


<!--/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////-->
<!--WHEEL_2-->
<!--Visual of the Link-->
<link name="wheel2">
    <visual>
      <geometry>
         <cylinder radius="${radius}" length="${length}"/>
      </geometry>
  <!-- Setting up the origin of the link -->
       <origin rpy="0 1.57075 1.57075" xyz="0 0 0"/>
       <material name="white"/>
    </visual>
     <collision>
<!--  COLLISION -->
     <origin rpy="0 1.57075 1.57075" xyz="0 0 0"/>
      <geometry>
         <cylinder radius="${radius}" length="${length}"/>
      </geometry>
    </collision>
<!--  INERTIAL -->
    <inertial>
      <origin xyz="0 0 0" rpy="0 1.5707 1.5707"/>
      <mass value="0.4"/>
      
      <inertia
        ixx="0.0006825" ixy="0.0" ixz="0.0"
        iyy="0.0006825" iyz="0.0"
        izz="0.001125"/>
    </inertial>
  </link>

<joint name="base_link_to_wheel2" type="continuous">
    <parent link="base_link"/>
    <child link="wheel2"/>
    <origin rpy="0 0 0" xyz="-0.125 0.23 0.075"/>
   <!--  <limit effort="100" velocity="100"/> -->
    <joint_properties damping="0.0" friction="0.0"/>
    <axis xyz="0 1 0"/>
   
  </joint>
<!--/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////-->
<!--WHEEL_3-->
<!--Visual of the Link-->
<link name="wheel3">
    <visual>
      <geometry>
  <!-- Setting up the origin of the link -->
         <cylinder radius="${radius}" length="${length}"/>
      </geometry>
       <origin rpy="0 1.57075 1.57075" xyz="0 0 0"/>
          <material name="white"/>
    </visual>
    <collision>
<!--  COLLISION -->
    <origin rpy="0 1.57075 1.57075" xyz="0 0 0"/>
      <geometry>
        <cylinder radius="${radius}" length="${length}"/>
      </geometry>
    </collision>
<!--  INERTIAL -->
    <inertial>
      <origin xyz="0 0 0" rpy="0 1.5707 1.5707"/>
      <mass value="0.4"/>
      
      <inertia
        ixx="0.0006825" ixy="0.0" ixz="0.0"
        iyy="0.0006825" iyz="0.0"
        izz="0.001125"/>
    </inertial>
  </link>

    <joint name="base_link_to_wheel3" type="continuous">
    <parent link="base_link"/>
    <child link="wheel3"/>
    <axis xyz="0 1 0"/>
   <!-- <limit effort="100" velocity="100"/>  -->
    <joint_properties damping="0.0" friction="0.0"/>
   <origin  xyz="0.125 -0.23 0.075"/>
     
  </joint>
<!--/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////-->
<!--WHEEL_4-->
<!--Visual of the Link-->
<link name="wheel4">
    <visual>
      <geometry>
  <!-- Setting up the origin of the link -->
        <cylinder radius="${radius}" length="${length}"/>
      </geometry>
       <origin rpy="0 1.57075 1.57075" xyz="0 0 0"/>
      <material name="white"/>
    </visual>
<collision>
<!--  COLLISION -->
<origin rpy="0 1.57075 1.57075" xyz="0 0 0"/>
      <geometry>
         <cylinder radius="${radius}" length="${length}"/>
      </geometry>
    </collision>
<!--  INERTIAL -->
    <inertial>
      <origin xyz="0 0 0" rpy="0 1.5707 1.5707"/>
      <mass value="0.4"/>
      
      <inertia
        ixx="0.0006825" ixy="0.0" ixz="0.0"
        iyy="0.0006825" iyz="0.0"
        izz="0.001125"/>
    </inertial>
  </link>

    <joint name="base_link_to_wheel4" type="continuous">
    <parent link="base_link"/>
    <child link="wheel4"/>
   <!-- <limit effort="100" velocity="100"/> -->
    <joint_properties damping="0.0" friction="0.0"/>
    <axis xyz="0 1 0"/>
    <origin  xyz="-0.125 -0.23 0.075"/>
    
  </joint>
<!--/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////-->


  <!--<link name="kinect">


    <visual>
      <geometry>
        <mesh filename="package://urdf_package/meshes/kinect.dae"/>
       </geometry>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      
    </visual>
    </link>

     <joint name="base_link_to_kinect" type="fixed">
  <axis xyz="0 0 0"/> 
<origin rpy="0 0 1.5707" xyz="0.30 0 0.5631"/>
 
    <parent link="base_link"/> 
    <child link="kinect"/>
        
   
  </joint> -->




<!--///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////-->
<!-- MANIPULATOR-->
<!-- BASE CYLINDER -->
<link name="base_cylinder">
    <visual>
    <geometry>
    <cylinder length="0.20" radius="0.09"/>
    </geometry>
    <origin rpy="0 0 0" xyz="0 0 0.10"/>
    <material name="blue"/>
    </visual>
   
    <collision>
    <geometry>
    <cylinder length="0.21" radius="0.10"/>
    </geometry>
    <origin rpy="0 0 0" xyz="0 0 0.10"/>
    </collision>

    <!-- PHYSICAL PROPERTIES -->
    <xacro:cylinder_inertial mass="3" length="0.20" radius="0.09" x="0" y="0" z="0.10"/>
</link>

    <joint name="base_link_to_base_cylinder" type="revolute">
    <parent link="base_link"/>
    <child link="base_cylinder"/>
    <origin rpy="0 0 0" xyz="0 0 0.5"/>
    <axis xyz="0 0 1"/>
    <limit effort="36" velocity="1" lower="-3.14" upper="3.14"/>
    <dynamics damping="50" friction="1"/>
    </joint>
<!--/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////-->
<!-- LINK 2 -->
<link name="link_2">
    <visual>
        <geometry>
        <box size="0.34 0.086 0.0795"/>
        </geometry>
        <origin rpy="0 1.57075 0" xyz="0.028 0 0.17"/>
    </visual>
    
    <collision>
        <geometry>
        <box size="0.35 0.086 0.0795"/>
        </geometry>
        <origin rpy="0 1.57075 0" xyz="0.028 0 0.17"/>
    </collision>

    <!-- PHYSICAL PROPERTIES -->
   
   <!--<box_inertial x="0.34" y="0.154" z="0.056" mass="1.153"/> -->

    <inertial>
    <mass value="3"/>
    <inertia ixx="0.030749" ixy="0.0" ixz="0.0" iyy="0.0304800625" iyz="0.0" izz="0.0034290625"/>
    <origin rpy="0 0 0" xyz="0 0 0.17"/>
    </inertial>

</link>

    <joint name="base_cylinder_to_link2" type="revolute">
    <parent link="base_cylinder"/>
    <child link="link_2"/>
    <origin rpy="0 0 0" xyz="0.09 0 0.20"/>
    <axis xyz="1 0 0"/>
    <limit effort="25" velocity="0.2094" lower="-1.57" upper="1.57"/>
    <dynamics damping="50" friction="1"/>
    </joint>
<!--///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////-->
<!-- LINK 3 -->
<link name="link_3">
    <visual>
        <geometry>
        <box size="0.265 0.03 0.04"/>
        </geometry>
        <origin rpy="0 1.57075 0" xyz="-0.032 0 0.1325"/>
        <material name="green"/>
    </visual>
    
    <collision>
        <geometry>
        <box size="0.275 0.03 0.074"/>
        </geometry>
        <origin rpy="0 1.57075 0" xyz="-0.032 0 0.1325"/>
    </collision>

    <!-- PHYSICAL PROPERTIES -->
    
    <!--<xacro:box_inertial x="0.30" y="0.1198" z="0.064" mass="0.744"/> -->
    
    <inertial>
    <mass value="0.120"/>
    <inertia ixx="0.00071125" ixy="0.0" ixz="0.0" iyy="0.00071825" iyz="0.0" izz="2.5e-05"/>
    <origin rpy="0 1.57075 0" xyz="-0.032 0 0"/>
    </inertial> 

</link>

    <joint name="link_2_to_link_3" type="revolute">
    <parent link="link_2"/>
    <child link="link_3"/>
    <origin rpy="0 0 0" xyz="0 0 0.34"/>
    <axis xyz="1 0 0"/>
    <limit effort="15" velocity="5.2359" lower="-1.57" upper="1.57"/>
    <dynamics damping="50" friction="1"/>
    </joint>

<!--//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////-->
<!-- MOTOR NO 4 -->
    <link name="motor_4">
    <visual>
        <geometry>
        <box size="0.07 0.07 0.07"/>
        </geometry>
        <origin rpy="0 0 0" xyz="-0.032 0 0"/>
        <material name="green"/>
    </visual>
    
    <collision>
        <geometry>
        <box size="0.08 0.08 0.08"/>
        </geometry>
        <origin rpy="0 1.57075 0" xyz="-0.032 0 0"/>
    </collision>
    
    <inertial>
    <mass value="0.624"/>
    <inertia ixx="0.0005096" ixy="0.0" ixz="0.0" iyy="0.0005096" iyz="0.0" izz="0.0005096"/>
    </inertial>
    </link>

    <joint name="motor_4_to_link_3" type="fixed">
    <parent link="link_3"/>
    <child link="motor_4"/>
    <origin rpy="0 0 0" xyz="0 0 0.30"/>
    </joint>

<!--///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////-->
<!-- LINK 4 -->
<link name="link_4">
    <visual>
        <geometry>
        <box size="0.09 0.0763 0.056"/>
        </geometry>
        <origin rpy="0 1.57075 0" xyz="0.028 0 0.045"/>
        <material name="blue"/>
    </visual>
    
    <collision>
        <geometry>
        <box size="0.10 0.0863 0.066"/>
        </geometry>
        <origin rpy="0 1.57075 0" xyz="0.028 0 0.045"/>
    </collision>

    <!-- PHYSICAL PROPERTIES -->
   
   <!--<xacro:box_inertial x="0.09" y="0.0763" z="0.056" mass="0.674"/> -->
   
    <inertial>
    <mass value="0.674"/>
    <inertia ixx="0.00050312" ixy="0.0" ixz="0.0" iyy="0.00063108" iyz="0.0" izz="0.00078193"/>
    <origin rpy="0 1.57075 0" xyz="0.028 0 0.045"/>
    </inertial>

</link>

    <joint name="link_3_to_link_4" type="revolute">
    <parent link="motor_4"/>
    <child link="link_4"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <axis xyz="1 0 0"/>
    <limit effort="5" velocity="3.14" lower="-1.57" upper="1.57"/>
    <dynamics damping="50" friction="1"/>
    </joint>
<!--///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////-->
<!-- GRIPPER BASE -->
<link name="gripper_base">
    <visual>
    <geometry>
    <cylinder length="0.07425" radius="0.055"/>
    </geometry>
    <origin rpy="0 0 0" xyz="0 0 0.037125"/>
    <material name="green"/>
    </visual>

    <collision>
    <geometry>
    <cylinder length="0.07425" radius="0.055"/>
    </geometry>
    <origin rpy="0 0 0" xyz="0 0 0.037125"/>
    </collision>

    <!-- PHYSICAL PROPERTIES -->

    <xacro:cylinder_inertial radius="0.055" length="0.07425" mass="0.3" x="0" y="0" z="0.03712"/>

</link>

    <joint name="link_4_to_gripper_base" type="revolute">
    <parent link="link_4"/>
    <child link="gripper_base"/>
    <origin rpy="0 0 0" xyz="0.028 0 0.09"/>
    <axis xyz="0 0 1"/>
    <limit effort="5" velocity="3.14" lower="-3.14" upper="3.14"/>
    <dynamics damping="50" friction="1"/>
    </joint>

<!--////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////-->
<link name="finger_base1">
    <visual>
        <geometry>
        <box size="0.062 0.03 0.03"/>
        </geometry>
        <origin rpy="0 1.57075 0" xyz="0 0 0.031"/>
    </visual>
    
    <collision>
        <geometry>
        <box size="0.062 0.03 0.03"/>
        </geometry>
        <origin rpy="0 1.57075 0" xyz="0 0 0.031"/>
    </collision>

    <!-- PHYSICAL PROPERTIES -->
   
  <!-- <xacro:box_inertial x="0.062" y="0.03" z="0.03" mass="0.025"/> -->

    <inertial>
    <mass value="0.025"/>
    <inertia ixx="9.8833333e-06" ixy="0.0" ixz="0.0" iyy="9.88333333333e-06" iyz="0.0" izz="3.75e-06"/>
    <origin rpy="0 0 0" xyz="0 0 0.031"/>
    </inertial>

</link>

    <joint name="gripper_base_to_finger_base1" type="revolute">
    <parent link="gripper_base"/>
    <child link="finger_base1"/>
    <origin rpy="0 0 0" xyz="0.055 0 0.07425"/>
    <axis xyz="0 1 0"/>
    <limit effort="30" velocity="1" lower="-0.0349066" upper="0.750492"/>
    <dynamics damping="50" friction="1"/>
    </joint>

<!--////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////-->
<link name="finger_top1">
    <visual>
        <geometry>
        <box size="0.052 0.028 0.016"/>
        </geometry>
        <origin rpy="0 1.57075 0" xyz="0 0 0.026"/>
    </visual>
    
    <collision>
        <geometry>
        <box size="0.052 0.028 0.016"/>
        </geometry>
        <origin rpy="0 1.57075 0" xyz="0 0 0.026"/>
    </collision>

    <!-- PHYSICAL PROPERTIES -->
  
   <!--<xacro:box_inertial x="0.052" y="0.028" z="0.016" mass="0.025"/> -->
  
    <inertial>
    <mass value="0.025"/>
    <inertia ixx="7.26666666667e-06" ixy="0.0" ixz="0.0" iyy="6.16666666667e-06" iyz="0.0" izz="2.16666666667e-06"/>
    <origin rpy="0 0 0" xyz="0 0 0.026"/>
    </inertial>
</link>

    <joint name="finger_base1_to_finger_top1" type="revolute">
    <parent link="finger_base1"/>
    <child link="finger_top1"/>
    <origin rpy="0 0 0" xyz="0 0 0.062"/>
    <axis xyz="0 1 0"/>
    <limit effort="30" velocity="1" lower="-0.785398" upper="0"/>
    <dynamics damping="50" friction="1"/>
    </joint>
<!--//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////-->
<link name="finger_base2">
    <visual>
        <geometry>
        <box size="0.062 0.03 0.03"/>
        </geometry>
        <origin rpy="0 1.57075 0" xyz="0 0 0.031"/>
    </visual>
    
    <collision>
        <geometry>
        <box size="0.062 0.03 0.03"/>
        </geometry>
        <origin rpy="0 1.57075 0" xyz="0 0 0.031"/>
    </collision>

    <!-- PHYSICAL PROPERTIES -->
   
   <!--<xacro:box_inertial x="0.062" y="0.03" z="0.03" mass="0.025"/> -->
   
  <inertial>
    <mass value="0.025"/>
    <inertia ixx="9.88333333333e-06" ixy="0.0" ixz="0.0" iyy="9.88333333333e-06" iyz="0.0" izz="3.75e-06"/>
    <origin rpy="0 0 0" xyz="0 0 0.031"/>
    </inertial>
</link>

    <joint name="gripper_base_to_finger_base2" type="revolute">
    <parent link="gripper_base"/>
    <child link="finger_base2"/>
    <origin rpy="0 0 2.0944" xyz=" -0.0275 0.047631 0.07425"/>
    <axis xyz="0 1 0"/>
    <limit effort="30" velocity="1" lower="-0.750492" upper="0.0349066"/>
    <dynamics damping="50" friction="1"/>
    </joint>

<!--////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////-->
<link name="finger_top2">
    <visual>
        <geometry>
        <box size="0.052 0.028 0.016"/>
        </geometry>
        <origin rpy="0 1.57075 0" xyz="0 0 0.026"/>
    </visual>
    
    <collision>
        <geometry>
        <box size="0.052 0.028 0.016"/>
        </geometry>
        <origin rpy="0 1.57075 0" xyz="0 0 0.026"/>
    </collision>

    <!-- PHYSICAL PROPERTIES -->
  
  <!--<xacro:box_inertial x="0.052" y="0.028" z="" mass="0.016"/>-->
  
   <inertial>
    <mass value="0.025"/>
    <inertia ixx="7.26666666667e-06" ixy="0.0" ixz="0.0" iyy="6.16666666667e-06" iyz="0.0" izz="2.16666666667e-06"/>
    <origin rpy="0 0 0" xyz="0 0 0.026"/>
    </inertial>
</link>

    <joint name="finger_base2_to_finger_top2" type="revolute">
    <parent link="finger_base2"/>
    <child link="finger_top2"/>
    <origin rpy="0 0 0" xyz="0 0 0.062"/>
    <axis xyz="0 1 0"/>
    <limit effort="30" velocity="1" lower="0" upper="-0.785398"/>
    <dynamics damping="50" friction="1"/>
    </joint>
<!--//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////-->
<link name="finger_base3">
    <visual>
        <geometry>
        <box size="0.062 0.03 0.03"/>
        </geometry>
        <origin rpy="0 1.57075 0" xyz="0 0 0.031"/>
    </visual>
    
    <collision>
        <geometry>
        <box size="0.062 0.03 0.03"/>
        </geometry>
        <origin rpy="0 1.57075 0" xyz="0 0 0.031"/>
    </collision>

    <!-- PHYSICAL PROPERTIES -->
  
  
  <inertial>
    <mass value="0.025"/>
    <inertia ixx="9.88333333333e-06" ixy="0.0" ixz="0.0" iyy="9.88333333333e-06" iyz="0.0" izz="3.75e-06"/>
    <origin rpy="0 0 0" xyz="0 0 0.031"/>
    </inertial>
</link>

    <joint name="gripper_base_to_finger_base3" type="revolute">
    <parent link="gripper_base"/>
    <child link="finger_base3"/>
    <origin rpy="0 0 0.872665" xyz=" -0.0275 -0.047631 0.07425"/>
    <axis xyz="0 1 0"/>
    <limit effort="30" velocity="1" lower="-0.0349066" upper="0.750492"/>
    <dynamics damping="50" friction="1"/>
    </joint>

<!--////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////-->
<link name="finger_top3">
    <visual>
        <geometry>
        <box size="0.052 0.028 0.016"/>
        </geometry>
        <origin rpy="0 1.57075 0" xyz="0 0 0.026"/>
    </visual>
    
    <collision>
        <geometry>
        <box size="0.052 0.028 0.016"/>
        </geometry>
        <origin rpy="0 1.57075 0" xyz="0 0 0.026"/>
    </collision>

    <!-- PHYSICAL PROPERTIES -->
   
   <!--<xacro:box_inertial x="0.052" y="0.028" z="0.016" mass="0.025"/>-->
   
   <inertial>
    <mass value="0.025"/>
    <inertia ixx="7.26666666667e-06" ixy="0.0" ixz="0.0" iyy="6.16666666667e-06" iyz="0.0" izz="2.16666666667e-06"/>
    <origin rpy="0 0 0" xyz="0 0 0.026"/>
    </inertial>
</link>

    <joint name="finger_base3_to_finger_top3" type="revolute">
    <parent link="finger_base3"/>
    <child link="finger_top3"/>
    <origin rpy="0 0 0" xyz="0 0 0.062"/>
    <axis xyz="0 1 0"/>
    <limit effort="30" velocity="1" lower="0" upper="0.785398"/>
    <dynamics damping="50" friction="1"/>
    </joint>
<!--/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////-->

<!--///////////////// TRANSMISSION TAGS FOR GAZEBO /////////////////////////////////-->
  <gazebo>
      <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
      <robotNamespace>/gazebo_MARS_prototype</robotNamespace>
      </plugin>
  </gazebo>

  
    <xacro:transmission_block joint_name="base_link_to_base_cylinder"/>
    <xacro:transmission_block joint_name="base_cylinder_to_link2"/>
    <xacro:transmission_block joint_name="link_2_to_link_3"/>
    <xacro:transmission_block joint_name="link_3_to_link_4"/>
    <xacro:transmission_block joint_name="link_4_to_gripper_base"/>

<!-- wheel 1 transmission tag-->


<transmission name="base_to_wheel1">
  <type>transmission_interface/SimpleTransmission</type>
  <actuator name="motor1">
  <mechanicalReduction>1</mechanicalReduction>
     <hardwareInterface>EffortJointInterface</hardwareInterface>

</actuator>
<joint name="base_link_to_wheel1">
   <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
</joint>
</transmission>


    <!-- /////////////////////////////////////////////////////////////////////-->
<!-- wheel 2 transmission tag-->

<transmission name="base_to_wheel2">
  <type>transmission_interface/SimpleTransmission</type>
  <actuator name="motor2">
  <mechanicalReduction>1</mechanicalReduction>
 <hardwareInterface>EffortJointInterface</hardwareInterface>

</actuator>
<joint name="base_link_to_wheel2">
   <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
</joint>
</transmission>

<!-- /////////////////////////////////////////////////////////////////////-->
<!-- wheel 3 transmission tag-->

<transmission name="base_to_wheel3">
  <type>transmission_interface/SimpleTransmission</type>
  <actuator name="motor3">
  <mechanicalReduction>1</mechanicalReduction>
   <hardwareInterface>EffortJointInterface</hardwareInterface>

</actuator>
<joint name="base_link_to_wheel3">
   <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
</joint>
</transmission>

<!-- /////////////////////////////////////////////////////////////////////-->
<!-- wheel 4 transmission tag-->

<transmission name="base_to_wheel4">
  <type>transmission_interface/SimpleTransmission</type>
  <actuator name="motor4">
  <mechanicalReduction>1</mechanicalReduction>
   <hardwareInterface>EffortJointInterface</hardwareInterface>

</actuator>
<joint name="base_link_to_wheel4">
   <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
</joint>
</transmission>

<!--/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////-->    
</robot>